### ISSUE TITLE
Cursor-based GraphQL pagination

### ISSUE THREAD

**Issue by dotansimha** [graphprotocol/graph-node#3113]
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the current behavior?**

Pagination is implemented on the root of every Query type, and returns a `ListType` of an entity.

**What is the expected behavior?**

We can implement Cursor-based pagination (see spec here https://relay.dev/graphql/connections.htm). It's supported in all popular clients, and makes pagination super easy and robust (since it's cursor based, so it's easier to get a reliable response, instead of using `skip`).

We can expose a `Connection` type on the root `Query`, without changing the existing - the new field can co-exists with the current API without breaking changes.

Here's an example: 

```graphql
type Query {
  purpose(id: ID): Purpose!
  purposes(filter: PurposeFilter): [Purpose!]!
  purposeConnection(filter: PurposeFilter, paginate: PaginationFilter): PurposeConnection!
}

input PaginationFilter {
  before: String
  after: String
  first: Int
  last: Int
}

type Purpose { ... }

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
  endCursor: String
}

type PurposeEdge {
  node: Purpose
  cursor: String!
}

type PurposeConnection {
  pageInfo: PageInfo!
  edges: [PurposeEdge]!
}
```

**Comment by dotansimha** [comment]
Closing, duplicate of: https://github.com/graphprotocol/graph-node/issues/613
